File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0;
        int r = heights.size()-1;
        int maxW = 0;
        while(l<r){
            int water = (r-l) * min(heights[l], heights[r]);
            maxW = max(water, maxW);
            if(heights[l]<=heights[r]){
                l++;
            }
            else if(heights[r]<=heights[l]){
                r--;
            }
        }
        return maxW;
    }
};

File: Data Structures & Algorithms/swim-in-rising-water/submission-0.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        adj = defaultdict(list)
        n = len(grid)
        visit = set()
        directions=[[-1,0], [1,0], [0,1], [0,-1]]
        #time/max-height, r, c
        minH = [[grid[0][0], 0, 0]]
        visit.add((0,0))
        while minH:
            t,r,c = heapq.heappop(minH)
            if r==n-1 and c==n-1:
                return t
            for dr,dc in directions:
                if min(r+dr, c+dc)<0 or r+dr==n or c+dc==n or(r+dr,c+dc) in visit:
                    continue
                visit.add((r+dr, c+dc))
                heapq.heappush(minH, [max(t, grid[r+dr][c+dc]), r+dr, c+dc])
            

